sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -46,7 +46,8 @@ import org.apache.spark.util.Utils
 class QueryExecution(
     val sparkSession: SparkSession,
     val logical: LogicalPlan,
-    val tracker: QueryPlanningTracker = new QueryPlanningTracker) {
+    val tracker: QueryPlanningTracker = new QueryPlanningTracker,
+    val alreadyOptimized: Boolean = false) {
```

Comment:
what about creating a subclass instead of adding this flag?

Reply:
I felt introducing an optional parameter was a bit simpler solution. Let's wait for some comments on this.